core/src/main/scala/io/snappydata/TableStatsProviderService.scala

```diff
@@ -0,0 +1,293 @@
+import com.gemstone.gemfire.cache.DataPolicy
+import com.gemstone.gemfire.cache.execute.FunctionService
+import com.gemstone.gemfire.i18n.LogWriterI18n
+import com.gemstone.gemfire.internal.SystemTimer
+import com.gemstone.gemfire.internal.cache.execute.InternalRegionFunctionContext
+import com.gemstone.gemfire.internal.cache.{PartitionedRegion, LocalRegion}
+import com.pivotal.gemfirexd.internal.engine.Misc
+import com.pivotal.gemfirexd.internal.engine.distributed.{GfxdMessage, GfxdListResultCollector}
+import com.pivotal.gemfirexd.internal.engine.distributed.GfxdListResultCollector.ListResultCollectorValue
+import com.pivotal.gemfirexd.internal.engine.sql.execute.MemberStatisticsMessage
+import com.pivotal.gemfirexd.internal.engine.store.{CompactCompositeKey, GemFireContainer}
+import com.pivotal.gemfirexd.internal.engine.ui.{SnappyRegionStatsCollectorFunction, SnappyRegionStatsCollectorResult, SnappyIndexStats, SnappyRegionStats}
+import com.pivotal.gemfirexd.internal.iapi.types.RowLocation
+import io.snappydata.Constant._
+
+import org.apache.spark.sql.catalyst.expressions.UnsafeRow
+import org.apache.spark.sql.collection.Utils
+import org.apache.spark.sql.execution.columnar.{JDBCSourceAsStore, JDBCAppendableRelation}
+import org.apache.spark.sql.execution.columnar.encoding.ColumnStatsSchema
+import org.apache.spark.unsafe.Platform
+import org.apache.spark.{SparkContext, Logging}
+import org.apache.spark.sql.{SnappySession, SnappyContext}
+
+trait TableStatsProviderService extends Logging {
+
+  @volatile
+  private var tableSizeInfo = Map.empty[String, SnappyRegionStats]
+  private val membersInfo = TrieMap.empty[String, mutable.Map[String, Any]]
+
+  private var _snc: Option[SnappyContext] = None
+
+  protected def snc: SnappyContext = synchronized {
+    _snc.getOrElse {
+      val context = SnappyContext()
+      _snc = Option(context)
+      context
+    }
+  }
+
+  @volatile protected var doRun: Boolean = false
+  @volatile private var running: Boolean = false
+
+
+  def start(sc: SparkContext): Unit = {
+    val delay = sc.getConf.getLong(Constant.SPARK_SNAPPY_PREFIX +
+        "calcTableSizeInterval", DEFAULT_CALC_TABLE_SIZE_SERVICE_INTERVAL)
+    doRun = true
+    Misc.getGemFireCache.getCCPTimer.schedule(
+      new SystemTimer.SystemTimerTask {
+        private val logger: LogWriterI18n = Misc.getGemFireCache.getLoggerI18n
+
+        override def run2(): Unit = {
+          try {
+            if (doRun) {
+              aggregateStats()
+            }
+          } catch {
+            case _: CancelException => // ignore
+            case e: Exception => if (!e.getMessage.contains(
+              "com.gemstone.gemfire.cache.CacheClosedException")) {
+              logger.warning(e)
+            } else {
+              logger.error(e)
+            }
+          }
+        }
+
+        override def getLoggerI18n: LogWriterI18n = {
+          logger
+        }
+      },
+      delay, delay)
+  }
```

Comment:
Why is this method in the trait and not in the embedded implementation? Can this ever be used by the thin implementation?